Cannot compile with GCC 10 or later for the CUDA backend if the kernel functor is templated

Looks like a change in GCC 10 breaks the compilation for the CUDA backend of kernels written with Alpaka, if the `operator()` is templated on its arguments other than `TAcc`.

Here is an extract of the code:
```c++
namespace cms {
  namespace alpakatools {

     struct countFromVector {
      template <typename T_Acc, typename Histo, typename T>
      ALPAKA_FN_ACC void operator()(const T_Acc &acc,
                                    Histo *__restrict__ h,
                                    uint32_t nh,
                                    T const *__restrict__ v,
                                    uint32_t const *__restrict__ offsets) const {
        ...
      }
    };

    template <typename Histo, typename T>
    ALPAKA_FN_HOST ALPAKA_FN_INLINE __attribute__((always_inline)) void fillManyFromVector(
        Histo *__restrict__ h,
        uint32_t nh,
        T const *__restrict__ v,
        uint32_t const *__restrict__ offsets,
        uint32_t totSize,
        unsigned int nthreads,
        ::ALPAKA_ACCELERATOR_NAMESPACE::Queue &queue) {
      launchZero(h, queue);

      ...

      alpaka::enqueue(queue,
                      alpaka::createTaskKernel<::ALPAKA_ACCELERATOR_NAMESPACE::Acc1D>(
                          workDiv, countFromVector(), h, nh, v, offsets));
      ...
    }

  }  // namespace alpakatools
}  // namespace cms
```

When used later on in some application code, _e.g._
```c++
::cms::alpakatools::fillManyFromVector(hits_d.phiBinner(), 10, hits_d.c_iphi(), hits_d.c_hitsLayerStart(), nHits, 256, queue);
```
this results in the error
```
.../external/alpaka/include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp: In instantiation of ‘static void alpaka::traits::Enqueue<alpaka::QueueUniformCudaHipRtNonBlocking, alpaka::TaskKernelGpuUniformCudaHipRt<TAcc, TDim, TIdx, TKernelFnObj, TArgs ...>, void>::enqueue(alpaka::QueueUniformCudaHipRtNonBlocking&, const alpaka::TaskKernelGpuUniformCudaHipRt<TAcc, TDim, TIdx, TKernelFnObj, TArgs ...>&) [with TAcc = alpaka::AccGpuCudaRt<std::integral_constant<long unsigned int, 1>, unsigned int>; TDim = std::integral_constant<long unsigned int, 1>; TIdx = unsigned int; TKernelFnObj = cms::alpakatools::countFromVector; TArgs = {cms::alpakatools::HistoContainer<short int, 128, 49152, 16, short unsigned int, 10>* __restrict__&, unsigned int&, const short int* __restrict__&, const unsigned int* __restrict__&}]’:
.../external/alpaka/include/alpaka/queue/Traits.hpp:48:58:   required from ‘void alpaka::enqueue(TQueue&, TTask&&) [with TQueue = alpaka::QueueUniformCudaHipRtNonBlocking; TTask = alpaka::TaskKernelGpuUniformCudaHipRt<alpaka::AccGpuCudaRt<std::integral_constant<long unsigned int, 1>, unsigned int>, std::integral_constant<long unsigned int, 1>, unsigned int, cms::alpakatools::countFromVector, cms::alpakatools::HistoContainer<short int, 128, 49152, 16, short unsigned int, 10>* __restrict__&, unsigned int&, const short int* __restrict__&, const unsigned int* __restrict__&>]’
.../src/alpaka/AlpakaCore/HistoContainer.h:113:16:   required from ‘void cms::alpakatools::fillManyFromVector(Histo*, uint32_t, const T*, const uint32_t*, uint32_t, unsigned int, alpaka_cuda_async::Queue&) [with Histo = cms::alpakatools::HistoContainer<short int, 128, 49152, 16, short unsigned int, 10>; T = short int; uint32_t = unsigned int; alpaka_cuda_async::Queue = alpaka::QueueUniformCudaHipRtNonBlocking]’
.../src/alpaka/plugin-SiPixelRecHits/alpaka/PixelRecHits.cc:72:125:   required from here
.../external/alpaka/include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:243:6: error: no matches converting function ‘uniformCudaHipKernel’ to type ‘void (*)(class alpaka::Vec<std::integral_constant<long unsigned int, 1>, unsigned int>, struct cms::alpakatools::countFromVector, class cms::alpakatools::HistoContainer<short int, 128, 49152, 16, short unsigned int, 10>*, unsigned int, const short int*, const unsigned int*)’
  243 |                 auto kernelName = uniform_cuda_hip::detail::
      |      ^         
.../external/alpaka/include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:74:1: note: candidate is: ‘template<class TAcc, class TDim, class TIdx, class TKernelFnObj, class ... TArgs> void alpaka::uniform_cuda_hip::detail::uniformCudaHipKernel(alpaka::Vec<TDim, TVal>, TKernelFnObj, TArgs ...)’
   74 |             __global__ void uniformCudaHipKernel(
      | ^           ~~~~~~~~
```

The errors happens with GCC 10.2 and 11.2; the same code compiles fine with GCC 9.

To reproduce:
```bash
git clone git@github.com:cms-patatrack/pixeltrack-standalone.git
cd pixeltrack-standalone
git checkout 2e46f69b77d8e0f4c1afd2b316e6b6b5dd884243
# edit the Makefile to point to a recent version of CUDA (default is /usr/local/cuda) and g++ from GCC 10
make alpaka
```

After having set up the environment and built the dependeicies, a simpler way to reproduce the problem is
```bash
nvcc -x cu  \
  -std=c++17  -O3  \
  -DALPAKA_ACC_GPU_CUDA_ENABLED  \
  -DALPAKA_ACC_GPU_CUDA_ONLY_MODE  \
  --expt-relaxed-constexpr --expt-extended-lambda  \
  -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored  \
  --diag-suppress 20014 \
  --cudart=shared  \
  -gencode arch=compute_75,code=[sm_75,compute_75]  \
  -ccbin g++-10  \
  --compiler-options '-fPIC -fdiagnostics-show-option -felide-constructors -fmessage-length=0 -fno-math-errno -ftree-vectorize -fvisibility-inlines-hidden --param vect-max-version-for-alias-checks=50 -msse3 -pipe -pthread'  \
  -DGPU_DEBUG  \
  -Isrc/alpaka  \
  -DSRC_DIR=src/alpaka  \
  -DLIB_DIR=lib/alpaka  \
  -isystem external/alpaka/include  \
  -isystem external/boost/include  \
  -isystem external/libbacktrace/include  \
  -dc \
  -c src/alpaka/test/alpaka/HistoContainer_t.cc  \
  -o HistoContainer_t.cc.cuda.o
```

I found a much simpler way to reproduce the problem, modifying the `vectorAdd` example to exhibit the same issue:
```diff
diff --git a/example/vectorAdd/src/vectorAdd.cpp b/example/vectorAdd/src/vectorAdd.cpp
index 3b5de083a72d..07dcc547f66a 100644
--- a/example/vectorAdd/src/vectorAdd.cpp
+++ b/example/vectorAdd/src/vectorAdd.cpp
@@ -163,12 +163,15 @@ auto main() -> int
     VectorAddKernel kernel;
 
     // Create the kernel execution task.
+    Data const* __restrict__ const pBufAccA(alpaka::getPtrNative(bufAccA));
+    Data const* __restrict__ const pBufAccB(alpaka::getPtrNative(bufAccB));
+    Data* __restrict__ const pBufAccC(alpaka::getPtrNative(bufAccC));
     auto const taskKernel = alpaka::createTaskKernel<Acc>(
         workDiv,
         kernel,
-        alpaka::getPtrNative(bufAccA),
-        alpaka::getPtrNative(bufAccB),
-        alpaka::getPtrNative(bufAccC),
+        pBufAccA,
+        pBufAccB,
+        pBufAccC,
         numElements);
 
     // Enqueue the kernel execution task
```

The modification compile fine with GCC 9:
```c++
$ g++ --version
g++ (GCC) 9.2.1 20191120 (Red Hat 9.2.1-2)
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ nvcc -std=c++17 -O3 -x cu -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored --expt-relaxed-constexpr --expt-extended-lambda --generate-line-info --source-in-ptx -Iinclude -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_ACC_GPU_CUDA_ONLY_MODE -DALPAKA_DEBUG=ALPAKA_DEBUG_FULL example/vectorAdd/src/vectorAdd.cpp -o vectorAdd
```
but fail with GCC 10:
```c++
$ g++ --version
g++ (GCC) 10.2.1 20201112 (Red Hat 10.2.1-8)
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ nvcc -std=c++17 -O3 -x cu -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored --expt-relaxed-constexpr --expt-extended-lambda --generate-line-info --source-in-ptx -Iinclude -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_ACC_GPU_CUDA_ONLY_MODE -DALPAKA_DEBUG=ALPAKA_DEBUG_FULL example/vectorAdd/src/vectorAdd.cpp -o vectorAdd
include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp: In instantiation of ‘static void alpaka::traits::Enqueue<alpaka::QueueUniformCudaHipRtBlocking, alpaka::TaskKernelGpuUniformCudaHipRt<TAcc, TDim, TIdx, TKernelFnObj, TArgs ...>, void>::enqueue(alpaka::QueueUniformCudaHipRtBlocking&, const alpaka::TaskKernelGpuUniformCudaHipRt<TAcc, TDim, TIdx, TKernelFnObj, TArgs ...>&) [with TAcc = alpaka::AccGpuCudaRt<std::integral_constant<long unsigned int, 1>, long unsigned int>; TDim = std::integral_constant<long unsigned int, 1>; TIdx = long unsigned int; TKernelFnObj = VectorAddKernel; TArgs = {const unsigned int* const __restrict__&, const unsigned int* const __restrict__&, unsigned int* const __restrict__&, const long unsigned int&}]’:
include/alpaka/queue/Traits.hpp:48:58:   required from ‘void alpaka::enqueue(TQueue&, TTask&&) [with TQueue = alpaka::QueueUniformCudaHipRtBlocking; TTask = const alpaka::TaskKernelGpuUniformCudaHipRt<alpaka::AccGpuCudaRt<std::integral_constant<long unsigned int, 1>, long unsigned int>, std::integral_constant<long unsigned int, 1>, long unsigned int, VectorAddKernel, const unsigned int* const __restrict__&, const unsigned int* const __restrict__&, unsigned int* const __restrict__&, const long unsigned int&>&]’
example/vectorAdd/src/vectorAdd.cpp:180:34:   required from here
include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:374:6: error: no matches converting function ‘uniformCudaHipKernel’ to type ‘void (*)(class alpaka::Vec<std::integral_constant<long unsigned int, 1>, long unsigned int>, struct VectorAddKernel, const unsigned int*, const unsigned int*, unsigned int*, long unsigned int)’
  374 |                 auto kernelName = uniform_cuda_hip::detail::
      |      ^         
include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:74:1: note: candidate is: ‘template<class TAcc, class TDim, class TIdx, class TKernelFnObj, class ... TArgs> void alpaka::uniform_cuda_hip::detail::uniformCudaHipKernel(alpaka::Vec<TDim, TVal>, TKernelFnObj, TArgs ...)’
   74 |             __global__ void uniformCudaHipKernel(
      | ^           ~~~~~~~~
```

--- a/include/alpaka/alpaka.hpp
+++ b/include/alpaka/alpaka.hpp
@@ -68,6 +68,7 @@
 #include <alpaka/core/Hip.hpp>
 #include <alpaka/core/OmpSchedule.hpp>
 #include <alpaka/core/Positioning.hpp>
+#include <alpaka/core/RemoveRestrict.hpp>
 #include <alpaka/core/Unroll.hpp>
 #include <alpaka/core/Unused.hpp>
 #include <alpaka/core/Utility.hpp>

--- a/include/alpaka/core/RemoveRestrict.hpp
+++ b/include/alpaka/core/RemoveRestrict.hpp
@@ -1,0 +1,40 @@
+/* Copyright 2021 Rene Widera
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/core/BoostPredef.hpp>
+
+namespace alpaka
+{
+    //! Removes __restrict__ from a type
+    template<typename T>
+    struct remove_restrict
+    {
+        using type = T;
+    };
+
+#if BOOST_COMP_MSVC
+    template<typename T>
+    struct remove_restrict<T* __restrict>
+    {
+        using type = T*;
+    };
+#else
+    template<typename T>
+    struct remove_restrict<T* __restrict__>
+    {
+        using type = T*;
+    };
+#endif
+
+    //! Helper to remove __restrict__ from a type
+    template<typename T>
+    using remove_restrict_t = typename remove_restrict<T>::type;
+} // namespace alpaka

--- a/include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp
+++ b/include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp
@@ -39,6 +39,7 @@
 // Implementation details.
 #    include <alpaka/acc/AccGpuUniformCudaHipRt.hpp>
 #    include <alpaka/core/Decay.hpp>
+#    include <alpaka/core/RemoveRestrict.hpp>
 #    include <alpaka/core/Unused.hpp>
 #    include <alpaka/dev/DevUniformCudaHipRt.hpp>
 #    include <alpaka/kernel/Traits.hpp>
@@ -141,7 +142,7 @@ namespace alpaka
         }
 
         TKernelFnObj m_kernelFnObj;
-        std::tuple<std::decay_t<TArgs>...> m_args;
+        std::tuple<remove_restrict_t<std::decay_t<TArgs>>...> m_args;
     };
 
     namespace traits
@@ -227,7 +228,7 @@ namespace alpaka
 
                 // Get the size of the block shared dynamic memory.
                 auto const blockSharedMemDynSizeBytes = meta::apply(
-                    [&](ALPAKA_DECAY_T(TArgs) const&... args) {
+                    [&](remove_restrict_t<ALPAKA_DECAY_T(TArgs)> const&... args) {
                         return getBlockSharedMemDynSizeBytes<TAcc>(
                             task.m_kernelFnObj,
                             blockThreadExtent,
@@ -242,7 +243,7 @@ namespace alpaka
                           << std::endl;
 #    endif
                 auto kernelName = uniform_cuda_hip::detail::
-                    uniformCudaHipKernel<TAcc, TDim, TIdx, TKernelFnObj, std::decay_t<TArgs>...>;
+                    uniformCudaHipKernel<TAcc, TDim, TIdx, TKernelFnObj, remove_restrict_t<std::decay_t<TArgs>>...>;
 
 #    if ALPAKA_DEBUG >= ALPAKA_DEBUG_FULL
 #        if defined(ALPAKA_ACC_GPU_CUDA_ENABLED)
@@ -268,7 +269,7 @@ namespace alpaka
                 // (MSVC). If not given by value, the kernel launch code does not copy the value but the pointer to the
                 // value location.
                 meta::apply(
-                    [&](ALPAKA_DECAY_T(TArgs) const&... args)
+                    [&](remove_restrict_t<ALPAKA_DECAY_T(TArgs)> const&... args)
                     {
 #    if defined(ALPAKA_ACC_GPU_CUDA_ENABLED)
                         kernelName<<<
@@ -357,7 +358,7 @@ namespace alpaka
 
                 // Get the size of the block shared dynamic memory.
                 auto const blockSharedMemDynSizeBytes = meta::apply(
-                    [&](ALPAKA_DECAY_T(TArgs) const&... args) {
+                    [&](remove_restrict_t<ALPAKA_DECAY_T(TArgs)> const&... args) {
                         return getBlockSharedMemDynSizeBytes<TAcc>(
                             task.m_kernelFnObj,
                             blockThreadExtent,
@@ -373,7 +374,7 @@ namespace alpaka
 #    endif
 
                 auto kernelName = uniform_cuda_hip::detail::
-                    uniformCudaHipKernel<TAcc, TDim, TIdx, TKernelFnObj, std::decay_t<TArgs>...>;
+                    uniformCudaHipKernel<TAcc, TDim, TIdx, TKernelFnObj, remove_restrict_t<std::decay_t<TArgs>>...>;
 #    if ALPAKA_DEBUG >= ALPAKA_DEBUG_FULL
                 // hipFuncAttributes not ported from HIP to HIP.
                 // TODO why this is currently not possible
@@ -396,7 +397,7 @@ namespace alpaka
 
                 // Enqueue the kernel execution.
                 meta::apply(
-                    [&](ALPAKA_DECAY_T(TArgs) const&... args)
+                    [&](remove_restrict_t<ALPAKA_DECAY_T(TArgs)> const&... args)
                     {
 #    if defined(ALPAKA_ACC_GPU_CUDA_ENABLED)
                         kernelName<<<

